### ISSUE TITLE
feat(model): candidate tree KV cache pruning and in-place compaction for speculative branch verification

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12725]
# feat(model): candidate tree KV cache pruning and in-place compaction for speculative branch verification

```routing
lane: internal/model
paths: internal/model/kv_tree_compaction.go, internal/model/kv_tree_compaction_test.go, internal/model/pagedkv.go, internal/model/speculative.go, internal/model/speculative_branch_cache.go, internal/model/speculative_contract.go, internal/model/speculative_compaction_test.go, internal/model/speculative_compaction_caps_test.go
expected_steps: 4
github_issue: 12725
```

## Lane

`internal/model` — inference engine, KV cache management, and speculative decoding coordination.

## Problem frame

- Centrality: Core
- Portfolio tier: 2 (serving)
- Priority: P1
- P1: advanced - eliminates full KV eviction and re-computation when unaccepted candidate tree branches are rejected
- P2: preserved - preserves bit-identical key and value activations for accepted tokens along the winning path
- P3: advanced - provides in-place compaction without heap reallocation across contiguous and paged KV layouts
- P4: preserved - integrates with SpeculativeBranchCache to allow opportunistic LRU branch activation recycling

## Parent context

Part of the Native Speculative Acceleration Series (`docs/tickets/native-speculative-acceleration/INDEX.md`), providing the memory compaction primitive required for candidate tree verification.

## Current state

Public commit `0f30c70c1cd2abfad8b8d08072943f196c80cb59` implements contiguous and paged tree KV compaction. The supported dense-cache tree verifier retains computed winning-path activations directly, preserves rejected branches before overwriting slots, and returns the actual terminal accepted logits. Invalid requests and unsupported recurrent/GLM/sparse side states retain conservative eviction/replay fallback.

Twelve focused tests pass in WSL, including exact synthetic dense-f32 sequential cache parity, an observed absence of accepted-token `Session.Step` replay, and retained branch-byte limits. The nine original tests also pass on physical hardware. The initial 2026-09-10 physical CPU run reports zero heap allocations but misses the <15 microsecond gate for the explicit 32-layer, 8-KV-head, 128-dimension fixture. The issue remains open. See the execution record below for source binding, evidence, and limits.

## Why this is next

Speculative tree decoding yields a much higher acceptance length per verification pass than linear chains, but without in-place KV tree pruning and compaction, the memory and computational overhead of re-populating accepted tokens destroys the latency benefit. A zero-allocation in-place compaction operator restores linear sequence contiguity in microseconds.

## Working spine

Target model accepts winning path indices `[]int{node_0, node_1, ...}` -> `KVTreePruner` optionally stages unaccepted branch activations into `SpeculativeBranchCache` -> in-place compaction shifts accepted key/value vectors into contiguous positions `[prefixLen, prefixLen+M)` -> cache length updates to `prefixLen + M` -> subsequent forward passes operate on a completely standard linear KV sequence without cache corruption.

## Core through-line

1. Define `KVTreePruner` and `PruneAndCompactTreeKV(cache *KVCache, prefixLen int, acceptedPath []int, treeSize int) error` in `internal/model/kv_tree_compaction.go`.
2. Implement in-place memory compaction for contiguous KV buffers (`[]float32` slices for K and V) by copying row vectors corresponding to `acceptedPath` into slots `prefixLen + 0, prefixLen + 1, ..., prefixLen + (M-1)` and setting the active length to `prefixLen + M`.
3. Support paged KV cache layouts (`PagedKVCache` in `internal/model/pagedkv.go`), re-indexing logical block tables and recycling physical pages assigned to unaccepted tree branches back to the freelist.
4. Wire unaccepted sibling branches into `SpeculativeBranchCache` (`internal/model/speculative_branch_cache.go`) prior to slot overwrites, allowing later activation reuse within configured retained-cache limits. Recycling latency is unmeasured.
5. Author comprehensive unit and stress tests verifying exact bit-for-bit key/value retention for accepted tokens, assertion of valid sequential position IDs, and zero memory leaks.

## Gold-plating boundary

Strictly excluded from this work unit:
- Garbage collection heuristics for arbitrary non-tree sparse attention graphs.
- Disk-spilling or host-RAM swapping of unaccepted candidate branches.
- Modifying target model weight tensors or quantization formats.

## Verifiable Witness

```bash
go test -v ./internal/model -run 'TestKVTreeCompaction' -count=1
```

## Witness

```bash
go test -v ./internal/model -run 'TestKVTreeCompaction' -count=1
```

### File:Line Seam

- `internal/model/kvcache.go:11 (KVCache)`
- `internal/model/kv_tree_compaction.go:10 (PruneAndCompactTreeKV)`
- `internal/model/kv_tree_compaction.go:86 (PruneAndCompactPagedTreeKV)`
- `internal/model/speculative.go:408 (ParallelVerifyKernel)`
- `internal/model/speculative_branch_cache.go:53 (CachedBranchActivation)`
- `internal/model/pagedkv.go:112 (PagedKVPool.alloc)`

### Blast Radius & Affected Lanes

- **Primary Lane:** `internal/model`.
- **Unaffected Lanes:** `internal/compute`, `internal/agent`, `platform/*` remain unaffected.

### Quarantined Fallback Mechanism

If tree compaction reports an invalid index or discontinuous path invariant, it safely falls back to conservative tail eviction back to `prefixLen` with full linear prefill of the accepted tokens, guaranteeing safety and correctness `[SW-VERIFIED]`.

## Scoped Acceptance Criteria

- [x] 1. [SW-VERIFIED] Implement `PruneAndCompactTreeKV` in `internal/model/kv_tree_compaction.go` supporting in-place compaction of contiguous KV caches.
- [x] 2. [SW-VERIFIED] Implement paged KV compaction in `internal/model/kv_tree_compaction.go` returning unaccepted tree pages to the paged block freelist.
- [x] 3. [SW-VERIFIED] Unit tests assert that accepted token KV states after compaction match non-speculative sequential generation down to exact bit parity (delta = 0.0).
- [x] 4. [SW-VERIFIED] Verify unaccepted branches are optionally captured by `SpeculativeBranchCache` without exceeding configured memory caps.
- [ ] 5. [HW-WITNESSED] Verify on physical hardware that KV compaction of a 32-node tree down to 4 accepted tokens completes in < 15 microseconds with zero heap allocations.

## Done condition

- [x] 1. [SW-VERIFIED] Implement `PruneAndCompactTreeKV` in `internal/model/kv_tree_compaction.go` supporting in-place compaction of contiguous KV caches.
- [x] 2. [SW-VERIFIED] Implement paged KV compaction in `internal/model/kv_tree_compaction.go` returning unaccepted tree pages to the paged block freelist.
- [x] 3. [SW-VERIFIED] Unit tests assert that accepted token KV states after compaction match non-speculative sequential generation down to exact bit parity (delta = 0.0).
- [x] 4. [SW-VERIFIED] Verify unaccepted branches are optionally captured by `SpeculativeBranchCache` without exceeding configured memory caps.
- [ ] 5. [HW-WITNESSED] Verify on physical hardware that KV compaction of a 32-node tree down to 4 accepted tokens completes in < 15 microseconds with zero heap allocations.

## Acceptance gate

`go test -v ./internal/model -run 'TestKVTreeCompaction' -count=1` exits 0 with 0 boundary violations.

## Closure binding

Resolving commit carries `(fak model)` trailer.

## Likely files

- `internal/model/kv_tree_compaction.go`
- `internal/model/kv_tree_compaction_test.go`
- `internal/model/speculative_branch_cache.go`
- `internal/model/pagedkv.go`
- `internal/model/speculative.go`
- `internal/model/speculative_contract.go`
- `internal/model/speculative_compaction_test.go`
- `internal/model/speculative_compaction_caps_test.go`

## Expected steps

4

## Execution record — 2026-09-10

Implementation: public `fak` commit `0f30c70c1cd2abfad8b8d08072943f196c80cb59`, pushed and independently read back from `origin/main`. The seven changed files passed targeted boundary checks with zero violations and staged-content leak auditing. Post-commit context provenance is `SEPARATED`.

The focused `TestKVTreeCompaction` suite passed nine top-level tests on WSL and physical hardware. The real verifier witness compares all retained K/Kraw/V planes, positions, and token lineage against sequential execution of a synthetic dense-f32 model. Replacing the new verifier with the original baseline through a Go overlay makes that witness fail with four observed `Session.Step` calls; the new verifier records zero. This is cache-state parity evidence, not a claim of end-to-end physical-model quality.

Follow-up `839f9937098722e0f8ba180e70f79c2c22aee348` adds three passing WSL tests for capture before compaction, eviction driven solely by `MaxBytes`, and immediate removal of oversized entries. Its scoped landing also included 21 unintended reversions. Corrective commit `9d3b97c3f50b708d536d2bd7b32b6cc9f0b3d60c` restores all 21 byte-for-byte to parent `f0af47ecd0a56eb5d5e769fdcc6f6e4eac124a05`; read-back confirms the net follow-up is exactly the new test file. These independently authored tests witness criterion 4's retained-cache budget. They do not claim a cap on transient entry construction and do not change the hardware benchmark's source or binary.

The full model package run has four pre-existing Qwen35 NaN/parity failures, all reproduced with original baseline sources at `2edf807363b453421b0412074d8519206de2d387`. Existing public issues #12433, #12038, and #11987 track them. Focused acceptance tests pass; the full package is not green.

Physical measurement, **2026-09-10 06:01:24–06:01:25 UTC**: AMD Ryzen AI Max+ 395, Linux 7.2.3-arch1-3, performance governor, CPU affinity 0. Go 1.26.6 linux/amd64 GOAMD64=v1, CGO enabled, no custom build tags. Fixture: 32 layers × 8 KV heads × 128 dimensions, prefix 128, accepted tree indices `[0, 5, 17, 31]`; three moved rows across K/Kraw/V copy 1,179,648 payload bytes. P50 **19.0695 µs** (95% bootstrap CI 19.037–19.110), P90 **19.129 µs** (CI 19.074–19.151), **0 B/op, 0 allocs/op**. N=10 run averages of 100 operations each; these are not per-operation tail percentiles. The one-timer control also misses the ceiling (P50 19.112 µs). The ticket does not define model/cache geometry, so the 15 µs hardware criterion remains unchecked.

Exact source/binary hashes, temperature/frequency snapshot, commands, raw observations, baseline failures, and confidence method are preserved in [`docs/benchmarks/receipts/kv-tree-prune-12725-20260910/README.md`](../../benchmarks/receipts/kv-tree-prune-12725-20260910/README.md) and [`receipt.json`](../../benchmarks/receipts/kv-tree-prune-12725-20260910/receipt.json). The benchmark binary was built from baseline plus staged tree `d073da63f267cc273acb3a9ec62a76078ed65d15` before the trunk merge; the seven changed source files match the landed commit exactly.

Zero allocation is scoped to contiguous and exclusive paged primitives. Shared destination pages may allocate for copy-on-write. Branch-cache limits govern retained entries after capture; transient construction and whole-engine allocations are separate. Device/GPU compaction and physical-model serving latency are unmeasured.
